File: python/memory_sync/encoding.py
"""Claude Code project-dir encoding (Step 0 verified, 2026-07-07)."""
import os
import re


def dashed(cwd: str) -> str:
    """Encode a cwd to Claude Code's project-dir name.

    Rule (verified): resolve realpath (macOS /tmp -> /private/tmp), then replace
    EVERY non-alphanumeric char (including '/' AND '_') with '-'. Case preserved.
    NOT just '/' -> '-'.
    """
    return re.sub(r"[^a-zA-Z0-9]", "-", os.path.realpath(cwd))


def project_dir(cwd: str) -> str:
    """The on-disk project directory path for a cwd."""
    return os.path.join(os.path.expanduser("~"), ".claude", "projects", dashed(cwd))

File: go/memorysync/encoding.go
package memorysync

import (
	"os"
	"path/filepath"
	"regexp"
)

var nonAlnum = regexp.MustCompile(`[^a-zA-Z0-9]`)

// Dashed encodes a cwd to Claude Code's project-dir name.
//
// Rule (Step 0 verified): resolve realpath (filepath.EvalSymlinks), then replace
// every non-alphanumeric char (including '/' AND '_') with '-'. Case preserved.
// If the path doesn't exist (e.g. synthetic test paths), EvalSymlinks errors and
// we fall back to the path as-is — tests are self-consistent (use Dashed() to
// compute expected, not hardcoded strings).
func Dashed(cwd string) string {
	real, err := filepath.EvalSymlinks(cwd)
	if err != nil {
		real = cwd
	}
	return nonAlnum.ReplaceAllString(real, "-")
}

// ProjectDir returns the on-disk project directory path for a cwd.
func ProjectDir(cwd string) string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".claude", "projects", Dashed(cwd))
}
